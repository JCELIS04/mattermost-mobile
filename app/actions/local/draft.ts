// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {DeviceEventEmitter, Image} from 'react-native';

import {Navigation, Screens} from '@app/constants';
import {goToScreen} from '@app/screens/navigation';
import {isTablet} from '@app/utils/helpers';
import DatabaseManager from '@database/manager';
import {getDraft} from '@queries/servers/drafts';
import {logError} from '@utils/log';

export const switchToGlobalDrafts = async () => {
    const isTablelDevice = isTablet();
    if (isTablelDevice) {
        DeviceEventEmitter.emit(Navigation.NAVIGATION_HOME, Screens.GLOBAL_DRAFTS);
    } else {
        goToScreen(Screens.GLOBAL_DRAFTS, '', {}, {topBar: {visible: false}});
    }
};

export async function updateDraftFile(serverUrl: string, channelId: string, rootId: string, file: FileInfo, prepareRecordsOnly = false) {
    try {
        const {database, operator} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
        const draft = await getDraft(database, channelId, rootId);
        if (!draft) {
            return {error: 'no draft'};
        }

        const i = draft.files.findIndex((v) => v.clientId === file.clientId);
        if (i === -1) {
            return {error: 'file not found'};
        }

        // We create a new list to make sure we re-render the draft input.
        const newFiles = [...draft.files];
        newFiles[i] = file;
        draft.prepareUpdate((d) => {
            d.files = newFiles;
            d.updateAt = Date.now();
        });

        if (!prepareRecordsOnly) {
            await operator.batchRecords([draft], 'updateDraftFile');
        }

        return {draft};
    } catch (error) {
        logError('Failed updateDraftFile', error);
        return {error};
    }
}

export async function removeDraftFile(serverUrl: string, channelId: string, rootId: string, clientId: string, prepareRecordsOnly = false) {
    try {
        const {database, operator} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
        const draft = await getDraft(database, channelId, rootId);
        if (!draft) {
            return {error: 'no draft'};
        }

        const i = draft.files.findIndex((v) => v.clientId === clientId);
        if (i === -1) {
            return {error: 'file not found'};
        }

        if (draft.files.length === 1 && !draft.message) {
            draft.prepareDestroyPermanently();
        } else {
            draft.prepareUpdate((d) => {
                d.files = draft.files.filter((v, index) => index !== i);
                d.updateAt = Date.now();
            });
        }

        if (!prepareRecordsOnly) {
            await operator.batchRecords([draft], 'removeDraftFile');
        }

        return {draft};
    } catch (error) {
        logError('Failed removeDraftFile', error);
        return {error};
    }
}

export async function updateDraftMessage(serverUrl: string, channelId: string, rootId: string, message: string, prepareRecordsOnly = false) {
    try {
        const {database, operator} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
        const draft = await getDraft(database, channelId, rootId);
        if (!draft) {
            if (!message) {
                return {};
            }

            const newDraft: Draft = {
                channel_id: channelId,
                root_id: rootId,
                message,
                update_at: Date.now(),
            };

            return operator.handleDraft({drafts: [newDraft], prepareRecordsOnly});
        }

        if (draft.message === message) {
            return {draft};
        }

        if (draft.files.length === 0 && !message) {
            draft.prepareDestroyPermanently();
        } else {
            draft.prepareUpdate((d) => {
                d.message = message;
                d.updateAt = Date.now();
            });
        }

        if (!prepareRecordsOnly) {
            await operator.batchRecords([draft], 'updateDraftMessage');
        }

        return {draft};
    } catch (error) {
        logError('Failed updateDraftMessage', error);
        return {error};
    }
}

export async function addFilesToDraft(serverUrl: string, channelId: string, rootId: string, files: FileInfo[], prepareRecordsOnly = false) {
    try {
        const {database, operator} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
        const draft = await getDraft(database, channelId, rootId);
        if (!draft) {
            const newDraft: Draft = {
                channel_id: channelId,
                root_id: rootId,
                files,
                message: '',
                update_at: Date.now(),
            };

            return operator.handleDraft({drafts: [newDraft], prepareRecordsOnly});
        }

        draft.prepareUpdate((d) => {
            d.files = [...draft.files, ...files];
            d.updateAt = Date.now();
        });

        if (!prepareRecordsOnly) {
            await operator.batchRecords([draft], 'addFilesToDraft');
        }

        return {draft};
    } catch (error) {
        logError('Failed addFilesToDraft', error);
        return {error};
    }
}

export const removeDraft = async (serverUrl: string, channelId: string, rootId = '') => {
    try {
        const {database} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
        const draft = await getDraft(database, channelId, rootId);
        if (draft) {
            await database.write(async () => {
                await draft.destroyPermanently();
            });
        }

        return {draft};
    } catch (error) {
        logError('Failed removeDraft', error);
        return {error};
    }
};

export async function updateDraftPriority(serverUrl: string, channelId: string, rootId: string, postPriority: PostPriority, prepareRecordsOnly = false) {
    try {
        const {database, operator} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
        const draft = await getDraft(database, channelId, rootId);
        if (!draft) {
            const newDraft: Draft = {
                channel_id: channelId,
                root_id: rootId,
                metadata: {
                    priority: postPriority,
                },
                update_at: Date.now(),
            };

            return operator.handleDraft({drafts: [newDraft], prepareRecordsOnly});
        }

        draft.prepareUpdate((d) => {
            d.metadata = {
                ...d.metadata,
                priority: postPriority,
            };
            d.updateAt = Date.now();
        });

        if (!prepareRecordsOnly) {
            await operator.batchRecords([draft], 'updateDraftPriority');
        }

        return {draft};
    } catch (error) {
        logError('Failed updateDraftPriority', error);
        return {error};
    }
}

export async function updateDraftMarkdownImageMetadata({
    serverUrl,
    channelId,
    rootId,
    imageMetadata,
    prepareRecordsOnly = false,
}: {
    serverUrl: string;
    channelId: string;
    rootId: string;
    imageMetadata: Dictionary<PostImage | undefined>;
    prepareRecordsOnly?: boolean;
}) {
    try {
        const {database, operator} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
        const draft = await getDraft(database, channelId, rootId);
        if (!draft) {
            const newDraft: Draft = {
                channel_id: channelId,
                root_id: rootId,
                metadata: {
                    images: imageMetadata,
                },
                update_at: Date.now(),
            };

            return operator.handleDraft({drafts: [newDraft], prepareRecordsOnly});
        }
        draft.prepareUpdate((d) => {
            d.metadata = {
                ...d.metadata,
                images: imageMetadata,
            };
            d.updateAt = Date.now();
        });
        if (!prepareRecordsOnly) {
            await operator.batchRecords([draft], 'updateDraftImageMetadata');
        }

        return {draft};
    } catch (error) {
        logError('Failed updateDraftImages', error);
        return {error};
    }
}

async function getImageMetadata(url: string) {
    let height = 0;
    let width = 0;
    let format;
    try {
        await new Promise((resolve, reject) => {
            Image.getSize(
                url,
                (imageWidth, imageHeight) => {
                    width = imageWidth;
                    height = imageHeight;
                    resolve(null);
                },
                (error) => {
                    logError('Failed to get image size', error);
                    reject(error);
                },
            );
        });
    } catch (error) {
        width = 0;
        height = 0;
    }
    const match = url.match(/\.(\w+)(?=\?|$)/);
    if (match) {
        format = match[1];
    }
    return {
        height,
        width,
        format,
        frame_count: 1,
    };
}

export async function parseMarkdownImages(markdown: string, imageMetadata: Dictionary<PostImage | undefined>) {
    let match;
    const imageRegex = /!\[.*?\]\((https:\/\/[^\s)]+)\)/g;
    while ((match = imageRegex.exec(markdown)) !== null) {
        const imageUrl = match[1];
        // eslint-disable-next-line no-await-in-loop
        imageMetadata[imageUrl] = await getImageMetadata(imageUrl);
    }
}
