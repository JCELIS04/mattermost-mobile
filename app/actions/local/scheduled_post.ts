// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {ActionType} from '@constants';
import DatabaseManager from '@database/manager';
import {logError} from '@utils/log';

export type ScheduledPostWebsocketEventPayload = {
    scheduled_post: string;
}

async function handleScheduledPosts(serverUrl: string, actionType: string, scheduledPosts: ScheduledPost[], prepareRecordsOnly = false) {
    if (!scheduledPosts.length) {
        return {models: undefined};
    }

    try {
        const {operator} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
        const models = await operator.handleScheduledPosts({actionType, scheduledPosts, prepareRecordsOnly});
        return {models};
    } catch (error) {
        logError('handleScheduledPosts cannot handle scheduled post websocket event', error);
        return {error};
    }
}

export async function handleCreateOrUpdateScheduledPost(serverUrl: string, msg: WebSocketMessage<ScheduledPostWebsocketEventPayload>, prepareRecordsOnly = false) {
    try {
        const scheduledPost: ScheduledPost = JSON.parse(msg.data.scheduled_post);
        return handleScheduledPosts(serverUrl, ActionType.SCHEDULED_POSTS.CREATE_OR_UPDATED_SCHEDULED_POST, [scheduledPost], prepareRecordsOnly);
    } catch (error) {
        logError('handleCreateOrUpdateScheduledPost cannot handle scheduled post added/update websocket event', error);
        return {error};
    }
}

export async function handleDeleteScheduledPost(serverUrl: string, msg: WebSocketMessage<ScheduledPostWebsocketEventPayload>, prepareRecordsOnly = false) {
    try {
        const scheduledPost: ScheduledPost = JSON.parse(msg.data.scheduled_post);
        return handleScheduledPosts(serverUrl, ActionType.SCHEDULED_POSTS.DELETE_SCHEDULED_POST, [scheduledPost], prepareRecordsOnly);
    } catch (error) {
        logError('handleDeleteScheduledPost cannot handle scheduled post deleted websocket event', error);
        return {error};
    }
}
