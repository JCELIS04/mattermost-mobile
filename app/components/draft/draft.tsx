// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {useIntl} from 'react-intl';
import {Keyboard, TouchableHighlight, View} from 'react-native';

import ChannelInfo from '@app/components/channel_info';
import DraftPost from '@app/components/draft/draft_post';
import {Screens} from '@app/constants';
import {useTheme} from '@app/context/theme';
import {useIsTablet} from '@app/hooks/device';
import {DRAFT_OPTIONS_BUTTON} from '@app/screens/draft_options';
import {openAsBottomSheet} from '@app/screens/navigation';
import {changeOpacity, makeStyleSheetFromTheme} from '@app/utils/theme';

import type ChannelModel from '@typings/database/models/servers/channel';
import type DraftModel from '@typings/database/models/servers/draft';
import type UserModel from '@typings/database/models/servers/user';

type Props = {
    channel: ChannelModel;
    location: string;
    draftReceiverUser?: UserModel;
    draft: DraftModel;
    layoutWidth: number;
}

const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
    return {
        container: {
            paddingHorizontal: 20,
            paddingVertical: 16,
            width: '100%',
            borderTopColor: changeOpacity(theme.centerChannelColor, 0.16),
            borderTopWidth: 1,
        },
        pressInContainer: {
            backgroundColor: changeOpacity(theme.centerChannelColor, 0.16),
        },
    };
});

const Draft: React.FC<Props> = ({
    channel,
    location,
    draft,
    draftReceiverUser,
    layoutWidth,
}) => {
    const intl = useIntl();
    const theme = useTheme();
    const style = getStyleSheet(theme);
    const isTablet = useIsTablet();

    const onLongPress = () => {
        Keyboard.dismiss();
        const title = isTablet ? intl.formatMessage({id: 'draft.options.title', defaultMessage: 'Draft Options'}) : 'Draft Options';
        openAsBottomSheet({
            closeButtonId: DRAFT_OPTIONS_BUTTON,
            screen: Screens.DRAFT_OPTIONS,
            theme,
            title,
            props: {channel, rootId: draft.rootId, draft},
        });
    };

    return (
        <TouchableHighlight
            delayLongPress={200}
            onLongPress={onLongPress}
            underlayColor={changeOpacity(theme.centerChannelColor, 0.1)}
        >
            <View
                style={[style.container]}
            >
                <ChannelInfo
                    channel={channel}
                    draftReceiverUser={draftReceiverUser}
                    rootId={draft.rootId}
                    testID='draft_post.channel_info'
                />
                <DraftPost
                    draft={draft}
                    location={location}
                    layoutWidth={layoutWidth}
                />
            </View>

        </TouchableHighlight>
    );
};

export default Draft;
