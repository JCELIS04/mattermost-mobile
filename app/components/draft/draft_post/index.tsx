// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {View, TouchableOpacity} from 'react-native';

import CompassIcon from '@app/components/compass_icon';
import DraftMessage from '@app/components/draft/draft_post/draft_message';
import {useTheme} from '@app/context/theme';
import {changeOpacity, makeStyleSheetFromTheme} from '@app/utils/theme';

import DraftFiles from './draft_files';

import type {DraftModel} from '@app/database/models/server';

type Props = {
    draft: DraftModel;
    location: string;
    layoutWidth: number;
}

const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
    return {
        container: {
            marginTop: 12,
        },
        acknowledgementContainer: {
            marginTop: 8,
            alignItems: 'center',
            borderRadius: 4,
            backgroundColor: changeOpacity(theme.onlineIndicator, 0.12),
            flexDirection: 'row',
            height: 32,
            width: 42,
            justifyContent: 'center',
            paddingHorizontal: 8,
        },
    };
});

const DraftPost: React.FC<Props> = ({
    draft,
    location,
    layoutWidth,
}) => {
    const theme = useTheme();
    const hasFiles = draft.files.length > 0;
    const acknowledgementsVisible = draft.metadata?.priority?.requested_ack;
    const style = getStyleSheet(theme);

    return (
        <View style={style.container}>
            <DraftMessage
                layoutWidth={layoutWidth}
                location={location}
                draft={draft}
            />
            {
                hasFiles &&
                <DraftFiles
                    filesInfo={draft.files}
                    isReplyPost={false}
                    draftId={draft.id}
                    location={location}
                    layoutWidth={layoutWidth}
                />
            }
            {
                acknowledgementsVisible &&
                <TouchableOpacity style={style.acknowledgementContainer}>
                    <CompassIcon
                        color={theme.onlineIndicator}
                        name='check-circle-outline'
                        size={24}
                    />
                </TouchableOpacity>
            }
        </View>
    );
};

export default DraftPost;
