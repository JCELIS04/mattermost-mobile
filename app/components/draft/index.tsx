// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {withDatabase, withObservables} from '@nozbe/watermelondb/react';
import {switchMap, of} from 'rxjs';

import {observeChannel, observeChannelMembers} from '@app/queries/servers/channel';
import {observeUser} from '@app/queries/servers/user';

import Drafts from './draft';

import type {WithDatabaseArgs} from '@typings/database/database';

type Props = {
    channelId: string;
} & WithDatabaseArgs;

const enhance = withObservables(['channelId'], ({channelId, database}: Props) => {
    const channel = observeChannel(database, channelId);
    const sendToUser = channel.pipe(
        switchMap((channelData) => {
            if (channelData?.type === 'D') {
                // Fetch the channel member for direct message channels
                return observeChannelMembers(database, channelId).pipe(
                    // eslint-disable-next-line max-nested-callbacks
                    switchMap((members) => {
                        if (members.length > 0) {
                            const userId = members[0]?.userId;
                            return observeUser(database, userId);
                        }
                        return of(undefined);
                    }),
                );
            }
            return of(undefined);
        }),
    );

    return {
        channel,
        sendToUser,
    };
});

export default withDatabase(enhance(Drafts));
