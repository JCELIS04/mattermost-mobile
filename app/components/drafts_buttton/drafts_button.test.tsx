// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {fireEvent} from '@testing-library/react-native';
import React from 'react';

import {switchToGlobalDrafts} from '@actions/local/draft';
import {renderWithIntl} from '@test/intl-test-helper';

import DraftsButton from './drafts_button';

jest.mock('@actions/local/draft', () => ({
    switchToGlobalDrafts: jest.fn(),
}));
describe('Drafts Button', () => {
    it('should render the drafts button component', () => {
        const wrapper = renderWithIntl(
            <DraftsButton
                draftsCount={1}
            />,
        );

        const {getByText} = wrapper;
        expect(getByText('Drafts')).toBeTruthy();
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should render the drafts button component with drafts count', () => {
        const wrapper = renderWithIntl(
            <DraftsButton
                draftsCount={23}
            />,
        );

        const {getByText} = wrapper;
        expect(getByText('Drafts')).toBeTruthy();
        expect(getByText('23')).toBeTruthy();
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('calls switchToGlobalDrafts when pressed', () => {
        const wrapper = renderWithIntl(
            <DraftsButton
                draftsCount={5}
            />,
        );
        const {getByTestId} = wrapper;
        fireEvent.press(getByTestId('channel_list.drafts.button'));
        expect(switchToGlobalDrafts).toHaveBeenCalledTimes(1);
        expect(wrapper.toJSON()).toMatchSnapshot();
    });
});
