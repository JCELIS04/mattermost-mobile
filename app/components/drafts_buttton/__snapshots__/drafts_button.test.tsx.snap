// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drafts Button calls switchToGlobalDrafts when pressed 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="channel_list.drafts.button"
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
        },
        {
          "paddingLeft": 18,
          "paddingRight": 20,
        },
        false,
        false,
        {
          "minHeight": 40,
        },
      ]
    }
  >
    <Icon
      name="pencil-outline"
      style={
        [
          {
            "color": "rgba(255,255,255,0.5)",
            "fontSize": 24,
            "marginRight": 12,
          },
          false,
        ]
      }
    />
    <Text
      style={
        [
          {
            "flex": 1,
          },
          {
            "color": "#ffffff",
            "fontFamily": "OpenSans",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
          false,
        ]
      }
    >
      Drafts
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <Icon
        color="#ffffff"
        name="pencil-outline"
        size={14}
        style={
          {
            "opacity": 0.56,
          }
        }
      />
      <Text
        style={
          {
            "color": "#ffffff",
            "fontFamily": "OpenSans-SemiBold",
            "fontSize": 12,
            "fontWeight": "600",
            "lineHeight": 16,
            "opacity": 0.64,
          }
        }
      >
        5
      </Text>
    </View>
  </View>
</View>
`;

exports[`Drafts Button should render the drafts button component 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="channel_list.drafts.button"
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
        },
        {
          "paddingLeft": 18,
          "paddingRight": 20,
        },
        false,
        false,
        {
          "minHeight": 40,
        },
      ]
    }
  >
    <Icon
      name="pencil-outline"
      style={
        [
          {
            "color": "rgba(255,255,255,0.5)",
            "fontSize": 24,
            "marginRight": 12,
          },
          false,
        ]
      }
    />
    <Text
      style={
        [
          {
            "flex": 1,
          },
          {
            "color": "#ffffff",
            "fontFamily": "OpenSans",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
          false,
        ]
      }
    >
      Drafts
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <Icon
        color="#ffffff"
        name="pencil-outline"
        size={14}
        style={
          {
            "opacity": 0.56,
          }
        }
      />
      <Text
        style={
          {
            "color": "#ffffff",
            "fontFamily": "OpenSans-SemiBold",
            "fontSize": 12,
            "fontWeight": "600",
            "lineHeight": 16,
            "opacity": 0.64,
          }
        }
      >
        1
      </Text>
    </View>
  </View>
</View>
`;

exports[`Drafts Button should render the drafts button component with drafts count 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
  testID="channel_list.drafts.button"
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
        },
        {
          "paddingLeft": 18,
          "paddingRight": 20,
        },
        false,
        false,
        {
          "minHeight": 40,
        },
      ]
    }
  >
    <Icon
      name="pencil-outline"
      style={
        [
          {
            "color": "rgba(255,255,255,0.5)",
            "fontSize": 24,
            "marginRight": 12,
          },
          false,
        ]
      }
    />
    <Text
      style={
        [
          {
            "flex": 1,
          },
          {
            "color": "#ffffff",
            "fontFamily": "OpenSans",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
          false,
        ]
      }
    >
      Drafts
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <Icon
        color="#ffffff"
        name="pencil-outline"
        size={14}
        style={
          {
            "opacity": 0.56,
          }
        }
      />
      <Text
        style={
          {
            "color": "#ffffff",
            "fontFamily": "OpenSans-SemiBold",
            "fontSize": 12,
            "fontWeight": "600",
            "lineHeight": 16,
            "opacity": 0.64,
          }
        }
      >
        23
      </Text>
    </View>
  </View>
</View>
`;
